File: bindings/Python/tests/test_new_features.py
"""
Test suite for newly added features: covMat, model(), and Optim class
"""
import pylibkriging as lk
import numpy as np
import pytest


class TestCovMat:
    """Test covariance matrix computation"""
    
    def test_covMat_basic(self):
        """Test basic covMat functionality"""
        np.random.seed(123)
        n = 20
        X = np.random.uniform(size=(n, 2))
        y = np.sin(X[:, 0]) + np.cos(X[:, 1])
        
        # Fit model
        k = lk.Kriging(y.reshape(-1, 1), X, "matern3_2")
        
        # Test covMat computation
        X1 = np.random.uniform(size=(5, 2))
        X2 = np.random.uniform(size=(10, 2))
        
        cov = k.covMat(X1, X2)
        
        # Check dimensions
        assert cov.shape == (5, 10), f"Expected shape (5, 10), got {cov.shape}"
        
        # Check symmetry when X1 == X2
        cov_sym = k.covMat(X1, X1)
        assert cov_sym.shape == (5, 5)
        assert np.allclose(cov_sym, cov_sym.T), "Covariance should be symmetric"
        
        # Covariance should be positive semi-definite
        eigenvals = np.linalg.eigvals(cov_sym)
        assert np.all(eigenvals >= -1e-10), "Covariance should be positive semi-definite"
    
    def test_covMat_all_classes(self):
        """Test covMat for all Kriging classes"""
        np.random.seed(456)
        n = 15
        X = np.random.uniform(size=(n, 1))
        y = np.sin(3 * X).flatten()
        noise = 0.01 * np.ones(n)
        
        X_test = np.random.uniform(size=(5, 1))
        
        # Test Kriging
        k1 = lk.Kriging(y.reshape(-1, 1), X, "gauss")
        cov1 = k1.covMat(X_test, X_test)
        assert cov1.shape == (5, 5)
        
        # Test NoiseKriging
        k2 = lk.NoiseKriging(y.reshape(-1, 1), noise.reshape(-1, 1), X, "gauss")
        cov2 = k2.covMat(X_test, X_test)
        assert cov2.shape == (5, 5)
        
        # Test NuggetKriging
        k3 = lk.NuggetKriging(y.reshape(-1, 1), X, "gauss")
        cov3 = k3.covMat(X_test, X_test)
        assert cov3.shape == (5, 5)


class TestModel:
    """Test model() method for parameter introspection"""
    
    def test_model_basic(self):
        """Test basic model() functionality"""
        np.random.seed(789)
        n = 10
        X = np.random.uniform(size=(n, 1))
        y = np.exp(X).flatten()
        
        k = lk.Kriging(y.reshape(-1, 1), X, "matern5_2", "linear", True, "BFGS", "LL")
        
        # Get model parameters
        params = k.model()
        
        # Check that all expected keys are present
        expected_keys = ['kernel', 'optim', 'objective', 'theta', 'is_theta_estim',
                        'sigma2', 'is_sigma2_estim', 'X', 'centerX', 'scaleX',
                        'y', 'centerY', 'scaleY', 'normalize', 'regmodel',
                        'beta', 'is_beta_estim', 'F', 'T', 'M', 'z']
        
        for key in expected_keys:
            assert key in params, f"Missing key: {key}"
        
        # Check types and values
        assert params['kernel'] == 'matern5_2'
        assert params['optim'] == 'BFGS'
        assert params['objective'] == 'LL'
        assert params['normalize'] == True
        assert params['regmodel'] == 'linear'
        
        # Check array shapes
        assert params['X'].shape == (n, 1)
        assert params['y'].shape == (n,)
        assert len(params['theta']) > 0
        assert len(params['beta']) > 0
    
    def test_model_noise_kriging(self):
        """Test model() for NoiseKriging includes noise field"""
        np.random.seed(321)
        n = 12
        X = np.random.uniform(size=(n, 2))
        y = np.sin(X[:, 0]) * np.cos(X[:, 1])
        noise = 0.05 * np.ones(n)
        
        k = lk.NoiseKriging(y.reshape(-1, 1), noise.reshape(-1, 1), X, "gauss")
        params = k.model()
        
        # NoiseKriging should have 'noise' field
        assert 'noise' in params, "NoiseKriging model should have 'noise' field"
        assert params['noise'].shape == (n,)
    
    def test_model_nugget_kriging(self):
        """Test model() for NuggetKriging includes nugget fields"""
        np.random.seed(654)
        n = 15
        X = np.random.uniform(size=(n, 1))
        y = X.flatten() ** 2
        
        k = lk.NuggetKriging(y.reshape(-1, 1), X, "matern3_2")
        params = k.model()
        
        # NuggetKriging should have 'nugget' and 'is_nugget_estim' fields
        assert 'nugget' in params, "NuggetKriging model should have 'nugget' field"
        assert 'is_nugget_estim' in params, "NuggetKriging model should have 'is_nugget_estim' field"
        assert isinstance(params['nugget'], (int, float))
        assert isinstance(params['is_nugget_estim'], bool)


class TestOptim:
    """Test Optim class static configuration methods"""
    
    def test_optim_reparametrization(self):
        """Test reparametrization settings"""
        # Save original state
        orig_state = lk.Optim.is_reparametrized()
        
        try:
            # Test setter and getter
            lk.Optim.use_reparametrize(True)
            assert lk.Optim.is_reparametrized() == True
            
            lk.Optim.use_reparametrize(False)
            assert lk.Optim.is_reparametrized() == False
        finally:
            # Restore original state
            lk.Optim.use_reparametrize(orig_state)
    
    def test_optim_theta_bounds(self):
        """Test theta bound factor settings"""
        # Save original values
        orig_lower = lk.Optim.get_theta_lower_factor()
        orig_upper = lk.Optim.get_theta_upper_factor()
        
        try:
            # Test lower factor
            lk.Optim.set_theta_lower_factor(0.05)
            assert abs(lk.Optim.get_theta_lower_factor() - 0.05) < 1e-10
            
            # Test upper factor
            lk.Optim.set_theta_upper_factor(15.0)
            assert abs(lk.Optim.get_theta_upper_factor() - 15.0) < 1e-10
        finally:
            # Restore original values
            lk.Optim.set_theta_lower_factor(orig_lower)
            lk.Optim.set_theta_upper_factor(orig_upper)
    
    def test_optim_variogram_bounds(self):
        """Test variogram bounds heuristic"""
        orig_state = lk.Optim.variogram_bounds_heuristic_used()
        
        try:
            lk.Optim.use_variogram_bounds_heuristic(True)
            assert lk.Optim.variogram_bounds_heuristic_used() == True
            
            lk.Optim.use_variogram_bounds_heuristic(False)
            assert lk.Optim.variogram_bounds_heuristic_used() == False
        finally:
            lk.Optim.use_variogram_bounds_heuristic(orig_state)
    
    def test_optim_log_level(self):
        """Test log level settings"""
        orig_level = lk.Optim.get_log_level()
        
        try:
            for level in [0, 1, 2, 3]:
                lk.Optim.set_log_level(level)
                assert lk.Optim.get_log_level() == level
        finally:
            lk.Optim.set_log_level(orig_level)
    
    def test_optim_max_iteration(self):
        """Test max iteration settings"""
        orig_max = lk.Optim.get_max_iteration()
        
        try:
            lk.Optim.set_max_iteration(500)
            assert lk.Optim.get_max_iteration() == 500
            
            lk.Optim.set_max_iteration(1000)
            assert lk.Optim.get_max_iteration() == 1000
        finally:
            lk.Optim.set_max_iteration(orig_max)
    
    def test_optim_tolerances(self):
        """Test tolerance settings"""
        orig_grad = lk.Optim.get_gradient_tolerance()
        orig_obj = lk.Optim.get_objective_rel_tolerance()
        
        try:
            lk.Optim.set_gradient_tolerance(1e-6)
            assert abs(lk.Optim.get_gradient_tolerance() - 1e-6) < 1e-15
            
            lk.Optim.set_objective_rel_tolerance(1e-8)
            assert abs(lk.Optim.get_objective_rel_tolerance() - 1e-8) < 1e-15
        finally:
            lk.Optim.set_gradient_tolerance(orig_grad)
            lk.Optim.set_objective_rel_tolerance(orig_obj)
    
    def test_optim_thread_settings(self):
        """Test thread configuration"""
        orig_delay = lk.Optim.get_thread_start_delay_ms()
        orig_pool = lk.Optim.get_thread_pool_size()
        
        try:
            lk.Optim.set_thread_start_delay_ms(20)
            assert lk.Optim.get_thread_start_delay_ms() == 20
            
            lk.Optim.set_thread_pool_size(4)
            assert lk.Optim.get_thread_pool_size() == 4
        finally:
            lk.Optim.set_thread_start_delay_ms(orig_delay)
            lk.Optim.set_thread_pool_size(orig_pool)
    
    def test_optim_affects_optimization(self):
        """Test that Optim settings actually affect model fitting"""
        np.random.seed(999)
        n = 25
        X = np.random.uniform(size=(n, 1))
        y = np.sin(5 * X).flatten()
        
        # Save original settings
        orig_max_iter = lk.Optim.get_max_iteration()
        
        try:
            # Fit with low iteration limit
            lk.Optim.set_max_iteration(5)
            k1 = lk.Kriging(y.reshape(-1, 1), X, "gauss")
            ll1 = k1.logLikelihood()
            
            # Fit with higher iteration limit
            lk.Optim.set_max_iteration(100)
            k2 = lk.Kriging(y.reshape(-1, 1), X, "gauss")
            ll2 = k2.logLikelihood()
            
            # More iterations should generally give better or equal log-likelihood
            # (may not be strictly greater due to convergence)
            print(f"LogLikelihood with 5 iters: {ll1}, with 100 iters: {ll2}")
            assert ll2 >= ll1 - 1e-6, "More iterations should not worsen the fit significantly"
            
        finally:
            lk.Optim.set_max_iteration(orig_max_iter)


class TestConsistency:
    """Test consistency across different Kriging classes"""
    
    def test_all_classes_have_covMat(self):
        """Ensure all Kriging classes have covMat method"""
        np.random.seed(111)
        n = 10
        X = np.random.uniform(size=(n, 1))
        y = X.flatten()
        noise = 0.01 * np.ones(n)
        
        k1 = lk.Kriging(y.reshape(-1, 1), X, "gauss")
        k2 = lk.NoiseKriging(y.reshape(-1, 1), noise.reshape(-1, 1), X, "gauss")
        k3 = lk.NuggetKriging(y.reshape(-1, 1), X, "gauss")
        
        X_test = np.random.uniform(size=(3, 1))
        
        assert hasattr(k1, 'covMat')
        assert hasattr(k2, 'covMat')
        assert hasattr(k3, 'covMat')
        
        # All should work
        cov1 = k1.covMat(X_test, X_test)
        cov2 = k2.covMat(X_test, X_test)
        cov3 = k3.covMat(X_test, X_test)
        
        assert cov1.shape == (3, 3)
        assert cov2.shape == (3, 3)
        assert cov3.shape == (3, 3)
    
    def test_all_classes_have_model(self):
        """Ensure all Kriging classes have model method"""
        np.random.seed(222)
        n = 10
        X = np.random.uniform(size=(n, 1))
        y = X.flatten()
        noise = 0.01 * np.ones(n)
        
        k1 = lk.Kriging(y.reshape(-1, 1), X, "gauss")
        k2 = lk.NoiseKriging(y.reshape(-1, 1), noise.reshape(-1, 1), X, "gauss")
        k3 = lk.NuggetKriging(y.reshape(-1, 1), X, "gauss")
        
        assert hasattr(k1, 'model')
        assert hasattr(k2, 'model')
        assert hasattr(k3, 'model')
        
        # All should return dicts
        m1 = k1.model()
        m2 = k2.model()
        m3 = k3.model()
        
        assert isinstance(m1, dict)
        assert isinstance(m2, dict)
        assert isinstance(m3, dict)
        
        # Check class-specific fields
        assert 'noise' in m2
        assert 'nugget' in m3
        assert 'is_nugget_estim' in m3


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: bindings/Octave/mlibkriging/tests/test_new_features.m
% Test suite for newly added features: covMat, model(), and Optim class
% Run this test with: octave test_new_features.m

function test_new_features()
    fprintf('=== Testing New Features ===\n\n');
    
    test_covMat_basic();
    test_covMat_all_classes();
    test_model_basic();
    test_model_noise_kriging();
    test_model_nugget_kriging();
    test_optim_reparametrization();
    test_optim_theta_bounds();
    test_optim_variogram_bounds();
    test_optim_log_level();
    test_optim_max_iteration();
    test_optim_tolerances();
    test_optim_thread_settings();
    test_all_classes_have_covMat();
    test_all_classes_have_model();
    
    fprintf('\n=== All Tests Passed! ===\n');
end


%% Test covMat functionality
function test_covMat_basic()
    fprintf('Test: covMat basic functionality...');
    
    rand('seed', 123);
    n = 20;
    X = rand(n, 2);
    y = sin(X(:, 1)) + cos(X(:, 2));
    
    % Fit model
    k = Kriging(y, X, 'matern3_2');
    
    % Test covMat computation
    X1 = rand(5, 2);
    X2 = rand(10, 2);
    
    cov = k.covMat(X1, X2);
    
    % Check dimensions
    assert(all(size(cov) == [5, 10]), 'Expected size [5, 10]');
    
    % Check symmetry when X1 == X2
    cov_sym = k.covMat(X1, X1);
    assert(all(size(cov_sym) == [5, 5]));
    assert(max(max(abs(cov_sym - cov_sym'))) < 1e-10, 'Covariance should be symmetric');
    
    % Covariance should be positive semi-definite
    eigenvals = eig(cov_sym);
    assert(all(eigenvals >= -1e-10), 'Covariance should be positive semi-definite');
    
    fprintf(' PASSED\n');
end


function test_covMat_all_classes()
    fprintf('Test: covMat for all Kriging classes...');
    
    rand('seed', 456);
    n = 15;
    X = rand(n, 1);
    y = sin(3 * X);
    noise = 0.01 * ones(n, 1);
    
    X_test = rand(5, 1);
    
    % Test Kriging
    k1 = Kriging(y, X, 'gauss');
    cov1 = k1.covMat(X_test, X_test);
    assert(all(size(cov1) == [5, 5]));
    
    % Test NoiseKriging
    k2 = NoiseKriging(y, noise, X, 'gauss');
    cov2 = k2.covMat(X_test, X_test);
    assert(all(size(cov2) == [5, 5]));
    
    % Test NuggetKriging
    k3 = NuggetKriging(y, X, 'gauss');
    cov3 = k3.covMat(X_test, X_test);
    assert(all(size(cov3) == [5, 5]));
    
    fprintf(' PASSED\n');
end


%% Test model() functionality
function test_model_basic()
    fprintf('Test: model() basic functionality...');
    
    rand('seed', 789);
    n = 10;
    X = rand(n, 1);
    y = exp(X);
    
    k = Kriging(y, X, 'matern5_2', 'linear', true, 'BFGS', 'LL');
    
    % Get model parameters
    params = k.model();
    
    % Check that all expected fields are present
    expected_fields = {'kernel', 'optim', 'objective', 'theta', 'is_theta_estim', ...
                       'sigma2', 'is_sigma2_estim', 'X', 'centerX', 'scaleX', ...
                       'y', 'centerY', 'scaleY', 'normalize', 'regmodel', ...
                       'beta', 'is_beta_estim', 'F', 'T', 'M', 'z'};
    
    for i = 1:length(expected_fields)
        assert(isfield(params, expected_fields{i}), ...
               sprintf('Missing field: %s', expected_fields{i}));
    end
    
    % Check types and values
    assert(strcmp(params.kernel, 'matern5_2'));
    assert(strcmp(params.optim, 'BFGS'));
    assert(strcmp(params.objective, 'LL'));
    assert(params.normalize == true);
    assert(strcmp(params.regmodel, 'linear'));
    
    % Check array shapes
    assert(all(size(params.X) == [n, 1]));
    assert(length(params.y) == n);
    assert(length(params.theta) > 0);
    assert(length(params.beta) > 0);
    
    fprintf(' PASSED\n');
end


function test_model_noise_kriging()
    fprintf('Test: model() for NoiseKriging...');
    
    rand('seed', 321);
    n = 12;
    X = rand(n, 2);
    y = sin(X(:, 1)) .* cos(X(:, 2));
    noise = 0.05 * ones(n, 1);
    
    k = NoiseKriging(y, noise, X, 'gauss');
    params = k.model();
    
    % NoiseKriging should have 'noise' field
    assert(isfield(params, 'noise'), 'NoiseKriging model should have noise field');
    assert(length(params.noise) == n);
    
    fprintf(' PASSED\n');
end


function test_model_nugget_kriging()
    fprintf('Test: model() for NuggetKriging...');
    
    rand('seed', 654);
    n = 15;
    X = rand(n, 1);
    y = X .^ 2;
    
    k = NuggetKriging(y, X, 'matern3_2');
    params = k.model();
    
    % NuggetKriging should have 'nugget' and 'is_nugget_estim' fields
    assert(isfield(params, 'nugget'), 'NuggetKriging model should have nugget field');
    assert(isfield(params, 'is_nugget_estim'), ...
           'NuggetKriging model should have is_nugget_estim field');
    assert(isnumeric(params.nugget));
    assert(islogical(params.is_nugget_estim));
    
    fprintf(' PASSED\n');
end


%% Test Optim class functionality
function test_optim_reparametrization()
    fprintf('Test: Optim reparametrization...');
    
    % Save original state
    orig_state = Optim.is_reparametrized();
    
    try
        % Test setter and getter
        Optim.use_reparametrize(true);
        assert(Optim.is_reparametrized() == true);
        
        Optim.use_reparametrize(false);
        assert(Optim.is_reparametrized() == false);
    catch err
        % Restore original state even on error
        Optim.use_reparametrize(orig_state);
        rethrow(err);
    end
    
    % Restore original state
    Optim.use_reparametrize(orig_state);
    
    fprintf(' PASSED\n');
end


function test_optim_theta_bounds()
    fprintf('Test: Optim theta bounds...');
    
    % Save original values
    orig_lower = Optim.get_theta_lower_factor();
    orig_upper = Optim.get_theta_upper_factor();
    
    try
        % Test lower factor
        Optim.set_theta_lower_factor(0.05);
        assert(abs(Optim.get_theta_lower_factor() - 0.05) < 1e-10);
        
        % Test upper factor
        Optim.set_theta_upper_factor(15.0);
        assert(abs(Optim.get_theta_upper_factor() - 15.0) < 1e-10);
    catch err
        Optim.set_theta_lower_factor(orig_lower);
        Optim.set_theta_upper_factor(orig_upper);
        rethrow(err);
    end
    
    % Restore original values
    Optim.set_theta_lower_factor(orig_lower);
    Optim.set_theta_upper_factor(orig_upper);
    
    fprintf(' PASSED\n');
end


function test_optim_variogram_bounds()
    fprintf('Test: Optim variogram bounds...');
    
    orig_state = Optim.variogram_bounds_heuristic_used();
    
    try
        Optim.use_variogram_bounds_heuristic(true);
        assert(Optim.variogram_bounds_heuristic_used() == true);
        
        Optim.use_variogram_bounds_heuristic(false);
        assert(Optim.variogram_bounds_heuristic_used() == false);
    catch err
        Optim.use_variogram_bounds_heuristic(orig_state);
        rethrow(err);
    end
    
    Optim.use_variogram_bounds_heuristic(orig_state);
    
    fprintf(' PASSED\n');
end


function test_optim_log_level()
    fprintf('Test: Optim log level...');
    
    orig_level = Optim.get_log_level();
    
    try
        for level = [0, 1, 2, 3]
            Optim.set_log_level(level);
            assert(Optim.get_log_level() == level);
        end
    catch err
        Optim.set_log_level(orig_level);
        rethrow(err);
    end
    
    Optim.set_log_level(orig_level);
    
    fprintf(' PASSED\n');
end


function test_optim_max_iteration()
    fprintf('Test: Optim max iteration...');
    
    orig_max = Optim.get_max_iteration();
    
    try
        Optim.set_max_iteration(500);
        assert(Optim.get_max_iteration() == 500);
        
        Optim.set_max_iteration(1000);
        assert(Optim.get_max_iteration() == 1000);
    catch err
        Optim.set_max_iteration(orig_max);
        rethrow(err);
    end
    
    Optim.set_max_iteration(orig_max);
    
    fprintf(' PASSED\n');
end


function test_optim_tolerances()
    fprintf('Test: Optim tolerances...');
    
    orig_grad = Optim.get_gradient_tolerance();
    orig_obj = Optim.get_objective_rel_tolerance();
    
    try
        Optim.set_gradient_tolerance(1e-6);
        assert(abs(Optim.get_gradient_tolerance() - 1e-6) < 1e-15);
        
        Optim.set_objective_rel_tolerance(1e-8);
        assert(abs(Optim.get_objective_rel_tolerance() - 1e-8) < 1e-15);
    catch err
        Optim.set_gradient_tolerance(orig_grad);
        Optim.set_objective_rel_tolerance(orig_obj);
        rethrow(err);
    end
    
    Optim.set_gradient_tolerance(orig_grad);
    Optim.set_objective_rel_tolerance(orig_obj);
    
    fprintf(' PASSED\n');
end


function test_optim_thread_settings()
    fprintf('Test: Optim thread settings...');
    
    orig_delay = Optim.get_thread_start_delay_ms();
    orig_pool = Optim.get_thread_pool_size();
    
    try
        Optim.set_thread_start_delay_ms(20);
        assert(Optim.get_thread_start_delay_ms() == 20);
        
        Optim.set_thread_pool_size(4);
        assert(Optim.get_thread_pool_size() == 4);
    catch err
        Optim.set_thread_start_delay_ms(orig_delay);
        Optim.set_thread_pool_size(orig_pool);
        rethrow(err);
    end
    
    Optim.set_thread_start_delay_ms(orig_delay);
    Optim.set_thread_pool_size(orig_pool);
    
    fprintf(' PASSED\n');
end


%% Test consistency across classes
function test_all_classes_have_covMat()
    fprintf('Test: All classes have covMat...');
    
    rand('seed', 111);
    n = 10;
    X = rand(n, 1);
    y = X;
    noise = 0.01 * ones(n, 1);
    
    k1 = Kriging(y, X, 'gauss');
    k2 = NoiseKriging(y, noise, X, 'gauss');
    k3 = NuggetKriging(y, X, 'gauss');
    
    X_test = rand(3, 1);
    
    % All should work
    cov1 = k1.covMat(X_test, X_test);
    cov2 = k2.covMat(X_test, X_test);
    cov3 = k3.covMat(X_test, X_test);
    
    assert(all(size(cov1) == [3, 3]));
    assert(all(size(cov2) == [3, 3]));
    assert(all(size(cov3) == [3, 3]));
    
    fprintf(' PASSED\n');
end


function test_all_classes_have_model()
    fprintf('Test: All classes have model...');
    
    rand('seed', 222);
    n = 10;
    X = rand(n, 1);
    y = X;
    noise = 0.01 * ones(n, 1);
    
    k1 = Kriging(y, X, 'gauss');
    k2 = NoiseKriging(y, noise, X, 'gauss');
    k3 = NuggetKriging(y, X, 'gauss');
    
    % All should return structs
    m1 = k1.model();
    m2 = k2.model();
    m3 = k3.model();
    
    assert(isstruct(m1));
    assert(isstruct(m2));
    assert(isstruct(m3));
    
    % Check class-specific fields
    assert(isfield(m2, 'noise'));
    assert(isfield(m3, 'nugget'));
    assert(isfield(m3, 'is_nugget_estim'));
    
    fprintf(' PASSED\n');
end


% Run all tests
test_new_features();
